cirq/linalg/transformations_test.py

```diff
@@ -393,3 +393,169 @@ def test_partial_trace_invalid_inputs():
     with pytest.raises(ValueError, match='2'):
         cirq.partial_trace(
             np.reshape(np.arange(2 * 2 * 2 * 2), (2,) * 4), [2])
+
+
+def test_keep_qubits():
+
+    bell00 = np.array([1, 0, 0, 1]) / np.sqrt(2)
+    plus_x = np.array([1, 1]) / np.sqrt(2)
+    state = np.kron(bell00, plus_x).reshape(2,2,2)
+    print("BELL", cirq.keep_qubits(state, [0, 1]))
+
+
+    a = np.arange(4) / np.linalg.norm(np.arange(4))
+    b = (np.arange(8) + 3) / np.linalg.norm(np.arange(8) + 3)
+    state = np.kron(a, b)
+    state = state.reshape(2,2,2,2,2)
+    print("A",np.abs(cirq.keep_qubits(a.reshape(2,2), [0, 1])))
+
+    np.testing.assert_almost_equal(
+        np.abs(cirq.keep_qubits(state, [0, 1])), a.reshape(2,2))
+    np.testing.assert_almost_equal(
+        np.abs(cirq.keep_qubits(state, [2, 3, 4])), b.reshape(2,2,2))
+
+    np.testing.assert_almost_equal(
+        np.abs(cirq.keep_qubits(state, [0, 1, 2, 3, 4])),
+        np.kron(a, b).reshape(2,2,2,2,2))
+
+
+# def test_keep_qubits():
+#     a = np.arange(4) / np.linalg.norm(np.arange(4))
+#     b = (np.arange(8) + 3) / np.linalg.norm(np.arange(8) + 3)
+#     c = (np.arange(16) + 1) / np.linalg.norm(np.arange(16) + 1)
+#     state = np.kron(np.kron(a, b), c).reshape((2,) * 9)
+#
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [0, 1])), a.reshape(2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [2, 3, 4])), b.reshape(2,2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [5, 6, 7, 8])), c.reshape(2,2,2,2))
+#
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [0, 1, 2, 3, 4])),
+#         np.kron(a, b).reshape(2,2,2,2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [0, 1, 5, 6, 7, 8])),
+#         np.kron(a, c).reshape(2,2,2,2,2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [2, 3, 4, 5, 6, 7, 8])),
+#         np.kron(b, c).reshape(2,2,2,2,2,2,2))
+
+
```

Comment:
Write a test that uses the `cirq.testing.random_superposition` method to make two wavefunctions, then `np.kron` them together with some swaps, then see if the factoring splits them up correctly.